core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -226,48 +383,203 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
+        }
+        buffer.rewind()
+        this
+      }
+    }
+  }
+
+  // We create this method to save memory copy operation. It reads from the original message set and directly
+  // write the converted messages into new message set buffer. Hence we don't need to allocate memory for each message
+  // during message conversion.
+  private def convertNonCompressedMessages(offsetCounter: AtomicLong,
+                                           compactedTopic: Boolean,
+                                           now: Long,
+                                           timestampType: TimestampType,
+                                           toMagicValue: Byte): ByteBufferMessageSet = {
+    // Get message count, shallow iterator is in-place
+    val sizeInBytesAfterConversion = shallowValidBytes + this.internalIterator(isShallow = true).foldLeft(0)(
+      (sizeDiff, messageAndOffset) => sizeDiff + Message.headerSizeDiff(messageAndOffset.message.magic, toMagicValue))
+    val newBuffer = ByteBuffer.allocate(sizeInBytesAfterConversion)
+    var newMessagePosition = 0
+    this.internalIterator(isShallow = true).foreach {messageAndOffset =>
+      val message = messageAndOffset.message
+      validateMessageKey(message, compactedTopic)
+      newBuffer.position(newMessagePosition)
+      // write offset.
+      newBuffer.putLong(offsetCounter.getAndIncrement)
+      // Write new message size
+      val newMessageSize = message.size + Message.headerSizeDiff(message.magic, toMagicValue)
+      newBuffer.putInt(newMessageSize)
+      // Create new message buffer
+      val newMessageBuffer = newBuffer.slice()
+      newMessageBuffer.limit(newMessageSize)
+      // Convert message
+      message.convertToBuffer(toMagicValue, newMessageBuffer, now, timestampType)
+
+      newMessagePosition += MessageSet.LogOverhead + newMessageSize
     }
+    newBuffer.rewind()
+    new ByteBufferMessageSet(newBuffer)
   }
 
+  private def validateNonCompressedMessagesAndAssignOffsetInPlace(offsetCounter: AtomicLong,
+                                                                  now: Long,
+                                                                  compactedTopic: Boolean,
+                                                                  timestampType: TimestampType,
+                                                                  timestampDiffMaxMs: Long): ByteBufferMessageSet = {
+    // do in-place validation and offset assignment
+    var messagePosition = 0
+    buffer.mark()
+    while (messagePosition < sizeInBytes - MessageSet.LogOverhead) {
+      buffer.position(messagePosition)
+      buffer.putLong(offsetCounter.getAndIncrement())
+      val messageSize = buffer.getInt()
+      val messageBuffer = buffer.slice()
+      messageBuffer.limit(messageSize)
+      val message = new Message(messageBuffer)
+      message.ensureValid()
```

Comment:
It doesn't seem that we need to valid the input message here since this is already done in Log,analyzeAndValidateMessageSet().